File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
def solution(tickets):
    
    
    def dfs(arr):
        # print(f'arr: {arr} ')
        if len(arr) == length + 1:
            answer.append(arr)
            return
        
        for i in range(len(tickets)):
            # check가 안되어있는 곳 중
            if not check[i] and arr[-1] == tickets[i][0]:
                check[i] = True
                dfs(arr + [tickets[i][1]])
                check[i] = False
                
        
    answer = []
    tickets = sorted(tickets, key=lambda x: (x[0], x[1]))
    length = len(tickets)
    check = [False] * len(tickets)
    
    for i in range(len(tickets)):
        if tickets[i][0] == "ICN":
            check[i] = True
            dfs([tickets[i][0], tickets[i][1]])
            check[i] = False
    
    return sorted(answer)[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;
typedef vector<string> vs;
typedef vector<vs> vvs;
typedef map<string, vector<tuple<string, bool>>> graph;
typedef map<string, bool> msb;

bool is_all_visited(graph routes) {
    for (auto &r : routes) {
        for (auto &d : r.second) {
            if (!get<1>(d)) {
                return false;
            }
        }
    }

    return true;
}

bool dfs(string from, graph &routes, vs &answer, int ticket_count) {
    if (answer.size() == ticket_count + 1) {
        return true;
    }

    for (auto &dest : routes[from]) {
        if (get<1>(dest)) {
            continue;
        }

        get<1>(dest) = true;
        answer.push_back(get<0>(dest));
        
        if (dfs(get<0>(dest), routes, answer, ticket_count)) {
            return true;
        }

        get<1>(dest) = false;
        answer.pop_back();
    }

    return false;
}
vs solution(vvs tickets) {
    vs answer;
    graph routes;
    
    for (auto &ticket : tickets) {
        routes[ticket[0]].push_back({ticket[1], false});
    }

    for (auto &g : routes) {
        sort(g.second.begin(), g.second.end());
    }

    answer.push_back("ICN"); 
    dfs("ICN", routes, answer, tickets.size());
    return answer;
}
